File: Demo/GtSdkDemo-Pods/GtSdkDemo/Controller/Advance/AdvanceViewController.m
//
//  AdvanceViewController.m
//  GtSdkDemo
//
//  Created by ak on 2020/03/20.
//  Copyright © 2019 Gexin Interactive (Beijing) Network Technology Co.,LTD. All rights reserved.
//
#import <WidgetKit/WidgetKit.h>
#import "AdvanceViewController.h"
#import "AliasViewController.h"
#import "AliasUnbindViewController.h"
#import "TagViewController.h"
#import "BadgeViewController.h"
#import "MessageViewController.h"
#import "LiveActivityViewController.h"
#import "GtSdkDemo-Swift.h"

@interface AdvanceViewController ()
@property (nonatomic, weak) IBOutlet UIView *bindAliasView;
@property (nonatomic, weak) IBOutlet UIView *unbindAliasView;
@property (nonatomic, weak) IBOutlet UIView *setTagView;
@property (nonatomic, weak) IBOutlet UIView *sendMessageView;
@property (nonatomic, weak) IBOutlet UIView *setBadgeView;
@property (nonatomic, weak) IBOutlet UIView *resetBadgeView;
@property (nonatomic, weak) IBOutlet UILabel *sdkStateLabel;
@property (nonatomic, weak) IBOutlet UISwitch *runBackgroundEnableSwitch;
@property (nonatomic, weak) IBOutlet UIScrollView *scrollView;
@property (nonatomic, weak) IBOutlet UIView *liveActivityView;
@property (nonatomic, weak) IBOutlet UIView *controlsView;

@end

@implementation AdvanceViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"高级功能";
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(updateStatusView) name:GTSdkStateNotification object:nil];
    [self addGestureRecognizer];
//    [self updateStatusView];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)updateStatusView {
    SdkStatus status = [GeTuiSdk status];
    switch (status) {
        case SdkStatusStarted:
            self.sdkStateLabel.text = @"启动";
            break;
        case SdkStatusStarting:
            self.sdkStateLabel.text = @"正在启动";
            break;
        default:
            self.sdkStateLabel.text = @"停止";
            break;
    }
}

- (void)addGestureRecognizer {
    [self.bindAliasView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(bindAliasClick)]];
    [self.unbindAliasView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(unBindAliasClick)]];
    [self.setTagView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(setTagClick)]];
    [self.sendMessageView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(sendMessageClick)]];
    [self.setBadgeView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(setBadgeClick)]];
    [self.resetBadgeView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(resetBadgeClick)]];
    [self.liveActivityView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(liveActivityClick)]];
    [self.controlsView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(controlsClick)]];
    
}

// MARK: Click Action

- (void)bindAliasClick {
    [self.navigationController pushViewController:[AliasViewController new] animated:NO];
}

- (void)unBindAliasClick {
    [self.navigationController pushViewController:[AliasUnbindViewController new] animated:NO];
}

- (void)setTagClick {
    [self.navigationController pushViewController:[TagViewController new] animated:NO];
}

- (void)sendMessageClick {
    [self.navigationController pushViewController:[MessageViewController new] animated:NO];
}

- (void)setBadgeClick {
    [self.navigationController pushViewController:[BadgeViewController new] animated:NO];
}

- (void)resetBadgeClick {
    [GeTuiSdk resetBadge];
    [[UIApplication sharedApplication] setApplicationIconBadgeNumber:0];
    [Utils AlertControllerWithTitle:@"复位角标成功" andMessage:@""];
}

- (IBAction)runBackgroundEnable:(id)sender {
    [GeTuiSdk runBackgroundEnable:self.runBackgroundEnableSwitch.on];
}

- (void)liveActivityClick {
#if !TARGET_OS_MACCATALYST
    [self.navigationController pushViewController:[LiveActivityViewController new] animated:NO];
#endif
}

- (void)controlsClick {
    if (@available(iOS 18.0, *)) {
        [ControlsCenterUtils getTokens:^(NSDictionary<NSString *,NSString *> * _Nonnull tokens) {
            [GeTuiSdk registerControlsTokens:tokens sequenceNum:@"sn1"];
            dispatch_async(dispatch_get_main_queue(), ^{
                [Utils AlertControllerWithTitle:@"注册Controls Token" andMessage:[tokens debugDescription]];
            });
        }];
    }
}

@end

File: Demo/GtSdkDemo-swift/GtSdkDemo-swift/Controller/Advance/AdvanceViewController.swift
//
//  AdvanceViewController.swift
//  GtSdkDemo-swift
//
//  Created by ak on 2020/03/20.
//  Copyright © 2020 Gexin Interactive (Beijing) Network Technology Co.,LTD. All rights reserved.
//

import UIKit
import GTSDK

class AdvanceViewController: UIViewController {
  
  @IBOutlet weak var bindAliasView: UIView!
  @IBOutlet weak var unbindAliasView: UIView!
  @IBOutlet weak var setTagView: UIView!
  @IBOutlet weak var sendMessageView: UIView!
  @IBOutlet weak var setBadgeView: UIView!
  @IBOutlet weak var resetBadgeView: UIView!
  @IBOutlet weak var runBackgroundEnableSwitch: UISwitch!
  @IBOutlet weak var scrollView: UIScrollView!
  @IBOutlet weak var sdkStateLabel: UILabel!
  @IBOutlet weak var liveActivityView: UIView!
  @IBOutlet weak var controlsView: UIView!
  
  
  override func viewDidLoad() {
    super.viewDidLoad()
    title = "高级功能"
    addGestureRecognizer()
    updateStatusView()
    NotificationCenter.default.addObserver(self, selector: #selector(updateStatusView), name: .GTSDKState, object: nil)
  }
  deinit {
    NotificationCenter.default.removeObserver(self)
  }
  
  @objc func updateStatusView() {
    sdkStateLabel.text = GeTuiSdk.status().title
  }
  
  func addGestureRecognizer() {
    bindAliasView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(bindAliasClick)))
    unbindAliasView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(unBindAliasClick)))
    setTagView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(setTagClick)))
    sendMessageView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(sendMessageClick)))
    setBadgeView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(setBadgeClick)))
    resetBadgeView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(resetBadgeClick)))
    liveActivityView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(liveActivityClick)))
    controlsView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(controlsClick)))
    
    
  }
  
  @objc func bindAliasClick() {
    navigationController?.pushViewController(AliasViewController(), animated: false)
  }
  
  @objc func unBindAliasClick() {
    navigationController?.pushViewController(AliasUnbindViewController(), animated: false)
  }
  
  @objc func setTagClick() {
    navigationController?.pushViewController(TagViewController(), animated: false)
  }
  
  @objc func sendMessageClick() {
    navigationController?.pushViewController(MessageViewController(), animated: false)
  }
  
  @objc func setBadgeClick() {
    navigationController?.pushViewController(BadgeViewController(), animated: false)
  }
  
  @objc func resetBadgeClick() {
    GeTuiSdk.resetBadge()
    UIApplication.shared.applicationIconBadgeNumber = 0
    UIAlertController.alert(title: "复位角标成功")
  }
  
  @IBAction func runBackgroundEnable(_ sender: Any) {
    GeTuiSdk.runBackgroundEnable(runBackgroundEnableSwitch.isOn)
  }
  
  @objc func liveActivityClick() {
    navigationController?.pushViewController(LiveActivityViewController(), animated: false)
  }
  
  @objc func controlsClick() {
    if #available(iOS 18.0, *) {
      ControlsCenterUtils.getTokens { tokens in
        GeTuiSdk.registerControlsTokens(tokens, sequenceNum: "sn1")
        DispatchQueue.main.async {
          UIAlertController.alert(title: "注册Controls Token", msg: tokens.description)
        }
      }
    }
  }
  
}
